Add support for authentication tokens

edufschmidt opened the pull request:
This PR updates `jellyfish-client-sdk` to work with authentication tokens.

Changes:
- Expect and handle `data.token.authentication` returned from API on session creation;
- Change `Authentication` header contents to contain the session ID and token separated by a delimiter in case both values are available. Otherwise, this header will only contain the session ID, in order to guarantee backwards compatibility.

All changes were applied in such a way that the SDK remains compatible with the current version of API, and should work even if only the session ID is returned.

Depends on:
- https://github.com/product-os/jellyfish/pull/7473
- https://github.com/product-os/jellyfish-action-library/pull/1377

joshbwlng commented:
@edufschmidt So this is a major change and therefore depends on the other authentication token PRs being merged first right? I believe the only project using this SDK is Jellyfish itself so that makes things a bit simpler, but we'll need to get the merge order correct.

joshbwlng commented:
Ah, just noticed your note about "PR set 1" and "PR set 2" in [the spec](https://jel.ly.fish/project-jellyfish-authentication-tokens-a39c917) so obviously you're aware of this. I'll go ahead and update this PRs description to make it more obvious here.

edufschmidt commented:
@joshbwlng If you agree, I would like to confirm first that we are OK with having two headers for passing auth credentials, as opposed to having a single one that contains both the session ID and the auth token separated by a delimiter, as this decision will impact the other related PRs as well.

joshbwlng commented:
@edufschmidt Responded in JF as well, but I feel like it would be cleaner to go with a single header instead of two. What do you think? I don't think we'd gain much by having two, and we can still be backwards compatible with a single header.

edufschmidt commented:
> @edufschmidt Responded in JF as well, but I feel like it would be cleaner to go with a single header instead of two. What do you think? I don't think we'd gain much by having two, and we can still be backwards compatible with a single header.

@joshbwlng I have combined both headers into a single one, and rebased my branch onto `master`. The remaining header has the format `Authorization: Bearer <sessionId>.<sessionToken>`. I also updated the `loginWithToken` method in the SDK to comply with these changes, and made sure the tests are passing.

Hades32 commented:
https://github.com/product-os/transformer-worker/tree/master/transformer-runner also uses the SDK and I think Hubot uses it, too

Hades32 commented:
Will there be a breaking change in the server API which needs session-id and token, soon? How do you get the token? Currently we're creating sessions for workers [via Fleet](https://github.com/product-os/transformers-fleet/blob/master/node-contractor/index.js#L205). How would we need to change that?

edufschmidt commented:
> Will there be a breaking change in the server API which needs session-id and token, soon? How do you get the token? Currently we're creating sessions for workers [via Fleet](https://github.com/product-os/transformers-fleet/blob/master/node-contractor/index.js#L205). How would we need to change that?

@Hades32, as pointed out int he PR description:

> The changes were applied in such a way that the SDK remains compatible with the current version of API, and should work even if only the session ID is returned.

The same thing applies to the server API so, in theory, there will be no need to change anything in the transformers. The server will work either if the `Authorization` header contains only `<sessionId>` (for compatibility reasons), or `<sessionId>.<sessionToken>`.

If you see any other potential impact this can have on transformers, or on any other module, please let us know. You can find more details about the spec on which this PR was based [here](https://jel.ly.fish/improvement-jip-authentication-tokens-9aea2bb0-2c76-4679-9062-6da2f54415c8@1.0.0/project-jellyfish-authentication-tokens-a39c917@1.0.0).

Hades32 commented:
so yes, this is outside this PR @edufschmidt, but according to your improvement (PR set 3), we will eventually require session tokens and therefore will need to change the Transformer registration flow to handle this. Do you have a timeline for this?

edufschmidt commented:
@Hades32 the PR set 3 has not yet been kickstarted, and we haven't yet defined a timeline for it, albeit the required changes should rather straightforward to implement. I believe the idea is to have PR sets 1 and 2 merged and then give some time for everyone to adapt to the new auth spec, but this should be further discussed in one of our arch calls.

Ereski commented on lib/auth.spec.ts:
```diff
@@ -0,0 +1,90 @@
+import Bluebird from 'bluebird';
```
Useful to have failure tests too, like info missing or in a bad header format

Ereski commented on lib/index.ts:
```diff
@@ -459,14 +530,24 @@ export class JellyfishSDK {
+			headers = {
+				authorization: `Bearer ${this.sessionId}`,
+			};
+		}
```
This logic is copied around multiple times. Would be better to factor it out in a separate method.

edufschmidt replied:
This was moved to a reusable method. Resolving this and all related threads.

Ereski commented on lib/index.ts:
```diff
@@ -378,15 +441,23 @@ export class JellyfishSDK {
+			headers = {
+				authorization: `Bearer ${this.sessionId}`,
+			};
+		}
```
One copy

Ereski commented on lib/index.ts:
```diff
@@ -618,13 +699,21 @@ export class JellyfishSDK {
+					authorization: `Bearer ${this.sessionId}`,
+				},
+			};
+		}
```
Another copy

Ereski commented on lib/index.ts:
```diff
@@ -647,13 +736,21 @@ export class JellyfishSDK {
+					authorization: `Bearer ${this.sessionId}`,
+				},
+			};
+		}
```
Another copy

Ereski commented on lib/index.ts:
```diff
@@ -684,13 +781,21 @@ export class JellyfishSDK {
+					authorization: `Bearer ${this.sessionId}`,
+				},
+			};
+		}
```
Another copy

Ereski reviewed (commented):
LGTM overall, but it's best for someone else to give approval as I'm not intimate with the SDK

edufschmidt commented:
@joshbwlng I think this PR is ready for merging. Could you please review it one last time?

LucianBuzzo commented on lib/auth.ts:
```diff
@@ -123,10 +127,11 @@ export class AuthSdk {
 			}
 
-			return token;
+			return sessionId;
```
This change makes the commit a breaking change. I also think this should possibly be returning an object containing both the session ID and the session token, I could be wrong about this though, it depends on what you think the implementing code should do

edufschmidt replied:
Previously, the SDK was referring to the session ID as `token` so, unless I'm missing something, this method is returning the exact same variable, just with a more descriptive (?) name. The only thing that changed was the method's signature, which now includes an optional argument `token`. If this argument is not provided, the `authToken` attribute is set to `null`, and the client defaults to using the session ID for authentication, thus ensuring compatibility. The idea was to eventually make `token` mandatory, and have everyone using the SDK adapt to this change. Does this make sense?

LucianBuzzo commented on lib/auth.spec.ts:
```diff
@@ -0,0 +1,90 @@
+	const mockSessionId = context.session;
+	const mockAuthToken = context.token;
+
+	server.post(/action/).reply((_uri) => {
```
I think that this will probably need to be changed as we recently switched to [dedicated endpoints for login](https://github.com/product-os/jellyfish-client-sdk/pull/528/files). Can you rebase this PR and fix anything that needs changes?

edufschmidt replied:
Yes, on it.

edufschmidt replied:
@LucianBuzzo the changes were applied, and the PR rebased. Please let me know if there is anything else you see that needs to be done.

edufschmidt commented:
@balena-ci rebase
